ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsLocator.java

```diff
@@ -68,28 +73,64 @@
   }
 
   @Override
-  public List<KsqlNode> locate(
-      final Struct key,
+  public List<KsqlLocation> locate(
+      final List<Struct> keys,
       final RoutingOptions routingOptions,
       final RoutingFilterFactory routingFilterFactory
   ) {
-    final KeyQueryMetadata metadata = kafkaStreams
-        .queryMetadataForKey(stateStoreName, key, keySerializer);
-
-    // Fail fast if Streams not ready. Let client handle it
-    if (metadata == KeyQueryMetadata.NOT_AVAILABLE) {
-      LOG.debug("KeyQueryMetadata not available for state store {} and key {}",
-                stateStoreName, key);
-      throw new MaterializationException(String.format(
-          "KeyQueryMetadata not available for state store %s and key %s", stateStoreName, key));
-    }
+    // Maintain request order for reproducibility by using a LinkedHashMap, even though it's
+    // not a guarantee of the API.
+    final Map<Integer, List<KsqlNode>> locationsByPartition = new LinkedHashMap<>();
+    final Map<Integer, Set<Struct>> keysByPartition = new HashMap<>();
+    final Set<Integer> filterPartitions = routingOptions.getPartitions();
+    for (Struct key : keys) {
```

Comment:
Can we make this loop cheaper for forwarded requests if we do send the partitions in the request?

Reply:
The only issue is that we still need to find which keys belong to which partitions to figure out which we should query for locally so we don't try them unnecessarily on rocksdb, which would also be expensive.  I could try to expose the info necessary to create a `DefaultStreamPartitioner`, but this seems a bit complex.  This may be a little wasteful, but it's not too bad.

Reply:
No it's ok. I also realized that there is not much to do that is a low hanging fruit. Just wanted to pick your brain if you had anything in mind. But it's not necessary to do now.